### ISSUE TITLE
Static Analysis

### ISSUE THREAD

**Issue by myiremark** [WorldHealthOrganization/app#338]

**Comment by myiremark** [comment]
https://pub.dev/packages/pedantic

**Comment by SamMousa** [comment]
@myiremark please elaborate a bit on what you mean with this issue!

**Comment by myiremark** [comment]
I'm not an expert in Dart/Pedantic, so I won't be able to perfectly elaborate on their differences.

We'd like to not just make sure the code is styled consistently and the way we want it written ala linter e.g Pedantic, we'd like to make sure it evaluates as intended securely and avoids common vulnerabilities at minimum.

It looks like Pedantic is actually built into Dart's static analysis tool in their SDK.

https://github.com/dart-lang/sdk/tree/master/pkg/analyzer

I can't speak to its abilities.

One suggestion I'm more familiar with was https://www.sonarqube.org/.

Dart is not immediately available as being a supported language.

Dart's analyzer may provide equivalent capabilities.

Perhaps a flutter expert can.expound?

**Comment by dnfield** [comment]
Pedantic provides a common set of analysis lints (code style rules) that are "recommended" for larger team projects.  It's also possible to just provide custom analysis rules in your `analysis_options.yaml`, but the advantage of pedantic is that people are likely familiar with it and it's easy to just drop in.

**Comment by dnfield** [comment]
And it's maintained by the Googler who manages what lints are applied on Google's Dart based projects.

**Comment by advayDev1** [comment]
I'm a fan of using an opinionated tool chosen by an expert over us spinning our own.

**Comment by SamMousa** [comment]
Already a PR open for pedantic btw

**Comment by suztomo** [comment]
I use this analysis rule "pedantic_mono" that derives from pedantic: https://pub.dev/packages/pedantic_mono 

> https://github.com/dart-lang/pedantic + https://github.com/flutter/samples

Its strict rules will provide consistent formatting among wide range of contributors.

**Comment by SamMousa** [comment]
It lacks a bit of documentation though; @suztomo could you summarize the reasoning behind those rules?

**Comment by suztomo** [comment]
The reasoning is that the more static analysis rules (that Google's Flutter team uses) more consistent the code will become.

While I don't care the reasoning behind each rule; for example, I don't believe constructors should come first ([sort_constructors_first](https://github.com/mono0926/pedantic_mono/blob/master/lib/analysis_options.yaml#L71)), I do think consistency matters, especially when there are many contributors with many backgrounds.

**Comment by SamMousa** [comment]
I follow your reasoning.

**Comment by suztomo** [comment]
Yes, aiming for this weekend. Note that my suggestion does not block onboarding to pedantic.

**Comment by SamMousa** [comment]
> Yes, aiming for this weekend. Note that my suggestion does not block onboarding to pedantic.

That's already been merged. Will leave this open for someone else to pick up, be sure to check back here before you start on your implementation!

Thanks!

**Comment by creativecreatorormaybenot** [comment]
I think we should just use pedantic as explained in #436. I can work on this today. It requries a bit more than adding the static analysis, i.e. fixing.

We should discuss if some rules should be excluded or not (if adding `pedantic` is agreed upon).

**Comment by SamMousa** [comment]
There is already a PR for basic pedantic, that part is good.
The addition would the rules in the mono package.

**Comment by SamMousa** [comment]
@dnfield would love your opinion on adding the rules from the mono package mentioned earlier.

**Comment by creativecreatorormaybenot** [comment]
@SamMousa I cannot seem to find that PR. 

I think that the rules in `mono` are somewhat arbitrarily chosen by the author. Considering, pedantic is maintained by the Dart team, it seems like a more reasonable choice. Custom rules can also be added manually.

**Comment by SamMousa** [comment]
> I think that the rules in `mono` are somewhat arbitrarily chosen by the author.

I agree

> Considering, pedantic is maintained by the Dart team, it seems like a more reasonable choice. Custom rules can also be added manually.

What's your recommendation? Sticking with default pedantic, with or without adding extra rules?
What do you think about the arguments given above?

**Comment by creativecreatorormaybenot** [comment]
I actually think that `pedantic` is already fairly strict and some rules like [`curly_braces_in_flow_control_structures`](https://dart-lang.github.io/linter/lints/curly_braces_in_flow_control_structures.html) are poorly implemented and should be excluded. 

There are also other extended rule sets, like [`extra_pedantic`](https://pub.dev/packages/extra_pedantic#-analysis-tab-), which is actually more popular than `pedantic_mono`, but they both have a low popularity score.

**Comment by SamMousa** [comment]
I'll discuss this during engineering standup, my recommendation based on your comments will be to use default pedantic for now. 

(This is always a bit of personal preference, so I'd rather stick with the standard)

Thanks for adding to the discussion!

**Comment by suztomo** [comment]
@creativecreatorormaybenot 

For my learning, would you be willing to explain the detail of this poor implementation? 

> curly_braces_in_flow_control_structures are poorly implemented

**Comment by creativecreatorormaybenot** [comment]
@suztomo Sure, I opened an issue about this https://github.com/dart-lang/site-www/issues/1994. This will probably be more insightful: https://github.com/dart-lang/linter/issues/1763

**Comment by creativecreatorormaybenot** [comment]
I realized now that the `pedantic` dependency was recently added: https://github.com/WorldHealthOrganization/app/pull/339

Any insight on why they are not using the latest version?

**Comment by dnfield** [comment]
It would probably be best to just add or remove lints as the project sees fit beyond pedantic.  I am not familiar with pedantic_mono.

**Comment by suztomo** [comment]
@creativecreatorormaybenot Thank you.

**Comment by creativecreatorormaybenot** [comment]
I fixed the `pedantic` version and the analysis in #448.

**Comment by britannio** [comment]
https://github.com/flutter/flutter/blob/master/analysis_options.yaml
These are the rules used in the Flutter repo, hopefully we can tighten up on code style soon by copying these in where appropriate.

**Comment by advayDev1** [comment]
Let's keep issues focused, updated title
